server/app/services/geo/esri/EsriClient.java

```diff
@@ -0,0 +1,198 @@
+package services.geo.esri;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.google.common.collect.ImmutableList;
+import com.typesafe.config.Config;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import javax.inject.Inject;
+import org.apache.commons.lang3.mutable.MutableInt;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import play.libs.Json;
+import play.libs.ws.WSBodyReadables;
+import play.libs.ws.WSBodyWritables;
+import play.libs.ws.WSClient;
+import play.libs.ws.WSRequest;
+import play.libs.ws.WSResponse;
+import services.Address;
+import services.AddressField;
+import services.geo.AddressLocation;
+import services.geo.AddressSuggestion;
+import services.geo.AddressSuggestionGroup;
+
+/**
+ * Provides methods for handling reqeusts to Esri geo services
+ *
+ * <p>@see <a
+ * href="https://developers.arcgis.com/rest/geocode/api-reference/geocoding-find-address-candidates.htm">Find
+ * Address Candidates</a>
+ *
+ * @see <a
+ *     href="https://developers.arcgis.com/rest/services-reference/enterprise/query-feature-service-layer-.htm">Query
+ *     (Map Service/Layer)</a>
+ */
+public class EsriClient implements WSBodyReadables, WSBodyWritables {
+  private final WSClient ws;
+
+  public static final String ESRI_CONTENT_TYPE = "application/json";
+  // Out Fields are passed to Esri geo services to indicate which fields should return in the
+  // response
+  public static final String ESRI_FIND_ADDRESS_CANDIDATES_OUT_FIELDS =
+      "Address, SubAddr, City, Region, Postal";
+  public static final String ESRI_FIND_ADDRESS_CANDIDATES_FORMAT = "json";
+  public Optional<String> ESRI_FIND_ADDRESS_CANDIDATES_URL;
+  public int ESRI_FIND_ADDRESS_CANDIDATES_TRIES;
+
+  private final Logger logger = LoggerFactory.getLogger(this.getClass());
+
+  @Inject
+  public EsriClient(Config configuration, WSClient ws) {
+    this.ws = checkNotNull(ws);
+    this.ESRI_FIND_ADDRESS_CANDIDATES_URL =
+        Optional.ofNullable(configuration.getString("esri_find_address_candidates_url"));
+    this.ESRI_FIND_ADDRESS_CANDIDATES_TRIES = Optional.ofNullable(configuration.getInt("esri_find_address_candidates_tries")).orElse(3);
+  }
+
+  /** Retries failed requests up to the provided value */
+  private CompletionStage<WSResponse> tryRequest(WSRequest request, MutableInt retries) {
+    CompletionStage<WSResponse> responsePromise = request.get();
+    responsePromise.handle(
+        (result, error) -> {
+          if (error != null || result.getStatus() != 200) {
+            logger.error("Esri API error: %s", error != null ? error.toString() : result.getStatusText());
```

Comment:
String formatting for [SL4J](https://www.slf4j.org/manual.html) works differently than `String.format`, unfortunately.

Needs to be:
```suggestion
            logger.error("Esri API error: {}", error != null ? error.toString() : result.getStatusText());
```